### PR TITLE
docs: Add v3 upgrading guide

### PR DESCRIPTION
**Pull request by mnapoli**
This is the "V3 upgrading guide" we had documented internally. I added that into the "Guides" documentation section.

cc @pgrzesik

### CODE REVIEW

File: docs/guides/upgrading-v3.md [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+That being said, some plugins need to be updated to be installable with v3. In most cases, it is a matter of allowing the plugin to be installed with Serverless Framework v3 in `package.json` (updating the `peerDependencies` requirement). Feel free to open an issue or pull request in the GitHub repository of the plugin.
+
+## Upgrading to v3
+
+First, upgrade to the latest v2 version and make sure that you do not see any deprecation warning.
+
+Then, to upgrade to Serverless Framework v3, run [the command that matches your installation method](../getting-started.md#upgrade):
+
+- MacOS/Linux: `serverless upgrade --major`
```
Comment (medikoo) [inline docs/guides/upgrading-v3.md:43]: I think this should clearly be indicated these method works only if user relies on _standalone_ installation.

I see a note to _run command that matches your installation method_, but I think many users may fast read. and just assume that they can magically upgrade on macOs with `serverless upgrade --major`, which will crash for majority of them (a small fraction of users, relies on standalone)

File: docs/guides/upgrading-v3.md [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+service:
+    name: my-service
+
+# Correct syntax:
+service: my-service
+```
+
+All options that used to be defined inside the `service` key have been moved to other sections (mentioned below in that document). This change clears up confusion that existed between the `service` and `provider` sections.
+
+CloudFormation extensions defined in `resources.extensions` are now validated: extensions that point to unknown resources will throw an error, which should help catch broken configuration before deploying ([learn more](../deprecations.md#defining-extensions-to-nonexistent-resources-in-resourcesextensions)).
```
Comment (medikoo) [inline docs/guides/upgrading-v3.md:106]: This paragraph can be removed, as this change was made in context of v2 (it was logical error to configure it as deprecation)

File: docs/guides/upgrading-v3.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        request:
+          # Deprecated in v2, removed in v3:
+          schema: ...
+          # Correct syntax:
+          schemas: ...
+```
+
+Learn more about [schema changes](../deprecations.md#aws-api-gateway-schemas).
+
+When an external API Gateway is used and imported via `provider.apiGateway.restApiId`, both `provider.logs.restApi` and `provider.tracing.apiGateway` options will be ignored. These settings are applicable only if the API Gateway is provisioned by Serverless Framework.
```
Comment (medikoo) [inline docs/guides/upgrading-v3.md:144]: Difference in v3 is that it'll throw (in v2 this was ignored)

File: docs/guides/upgrading-v3.md [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+
+The CloudFormation tags defined in `provider.tags` will now be correctly applied to HTTP APIs ([learn more](../deprecations.md#http-api-provider-tags)).
+
+Starting with v3, a bug fix for API Gateway authorizers will become the default behavior (to match AWS’ recommended behavior): `functions[].events[].http.authorizer.identitySource` will no longer be set to `method.request.header.Authorization` by default when caching is disabled (i.e. for authorizers of type "request" with `resultTtlInSeconds` set to "0") ([learn more](../deprecations.md#default-identitysource-for-httpauthorizer)).
```
Comment (medikoo) [inline docs/guides/upgrading-v3.md:148]: I would reword it to:

_Starting with v3 `functions[].events[].http.authorizer.identitySource` (setting related to API Gateway authorizers) will no longer be set to `method.request.header.Authorization` by default when caching is disabled_.

At least stating that _a bug fix for API Gateway authorizers will become the default behavior_ feels a bit weird to me

File: docs/guides/upgrading-v3.md [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+### CloudFront
+
+CloudFront’s `behavior` option (ForwardedValues, MinTTL, MaxTTL and DefaultTTL) has been removed, as this feature was deprecated by AWS. Use the new "cache policy" feature instead ([learn more](../deprecations.md#cloudfront-event-behaviorforwardedvalues-property)).
```
Comment (medikoo) [inline docs/guides/upgrading-v3.md:152]: Wouldn't it be nicer to link `new "cache policy feature instead` to AWS documentation resource,  instead of adding extra "learn more" link?

**Review by medikoo — COMMENTED** [review]
@mnapoli really great doc! I have just a few suggestions

File: docs/guides/upgrading-v3.md [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+The `schema` option on HTTP events has also been renamed to `schemas`. That change allowed us to support much more schema validation features:
+
+```yaml
+functions:
+  hello:
+    handler: hello.handler
+    events:
+      - http:
+        ...
+        request:
```
Comment (pgrzesik) [inline docs/guides/upgrading-v3.md:135]: nit: this should be intended one level deeper

File: docs/guides/upgrading-v3.md [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+          schema: ...
+          # Correct syntax:
+          schemas: ...
+```
+
+Learn more about [schema changes](../deprecations.md#aws-api-gateway-schemas).
+
+When an external API Gateway is used and imported via `provider.apiGateway.restApiId`, both `provider.logs.restApi` and `provider.tracing.apiGateway` options will be ignored. These settings are applicable only if the API Gateway is provisioned by Serverless Framework.
+
+The CloudFormation tags defined in `provider.tags` will now be correctly applied to HTTP APIs ([learn more](../deprecations.md#http-api-provider-tags)).
```
Comment (pgrzesik) [inline docs/guides/upgrading-v3.md:146]: Maybe let's add the word `stage` at the end? Because the tags are applied to HTTP API stage only

File: docs/guides/upgrading-v3.md (hunk 4 above)
Comment (pgrzesik) [reply docs/guides/upgrading-v3.md:148]: We should also mention that it only applies to authorizers of `request` type with caching disabled, not in all cases when caching is disabled

File: docs/guides/upgrading-v3.md (hunk 5 above)
Comment (pgrzesik) [reply docs/guides/upgrading-v3.md:152]: I think the current wording might be a bit confusing, as we still support `behavior` setting, just a few of it's options has been removed (the ones mentioned in parentheses).

**Review by pgrzesik — COMMENTED** [review]
Great document :clap: I only have few small suggestions, please let me know what do you think :bow:

File: docs/guides/upgrading-v3.md (hunk 1 above)
Comment (mnapoli) [reply docs/guides/upgrading-v3.md:43]: This is based on this documentation: https://www.serverless.com/framework/docs/getting-started#upgrade

How can we improve both? What can I explain to users so that they know what to do?

Comment (medikoo) [reply docs/guides/upgrading-v3.md:43]: Documentation on_standalone_ installation will probably work better as distinct page or section, and not to be mixed with _npm_ installation, which doesn't need much documentation aside of basic install command (it's in scope of npm package manager to document its commands)

File: docs/guides/upgrading-v3.md (hunk 4 above)
Comment (mnapoli) [reply docs/guides/upgrading-v3.md:148]: I've tried to push a change that still includes an explanation as to "why" the change was made (and removed the word "bugfix".

I also kept the details about "caching disabled".

Let me know if that's still correct.

File: docs/guides/upgrading-v3.md (hunk 5 above)
Comment (mnapoli) [reply docs/guides/upgrading-v3.md]: Good point, I added the link and rephrased it. Let me know if that's good.

File: docs/guides/upgrading-v3.md (hunk 1 above)
Comment (mnapoli) [reply docs/guides/upgrading-v3.md:43]: Right but how can we explain that to users?

I.e. as a user, reading the upgrade guide, what should I do? How do I know what instruction to follow?

**Review by pgrzesik — APPROVED** [review]
Thanks, the comments I've raised seem to be addressed so it's a green light from me :+1:

File: docs/guides/upgrading-v3.md [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+            schemas: ...
+```
+
+Learn more about [schema changes](../deprecations.md#aws-api-gateway-schemas).
+
+When an external API Gateway is used and imported via `provider.apiGateway.restApiId`, both `provider.logs.restApi` and `provider.tracing.apiGateway` options are ignored. In v3, an error will be thrown if these options are defined. Indeed, these settings are applicable only if the API Gateway is provisioned by Serverless Framework.
+
+The CloudFormation tags defined in `provider.tags` will now be correctly applied to HTTP APIs stages ([learn more](../deprecations.md#http-api-provider-tags)).
+
+Starting with v3, AWS’ recommended behavior for API Gateway authorizers will become the default: `functions[].events[].http.authorizer.identitySource` will no longer be set to `method.request.header.Authorization` by default when caching is disabled (i.e. for authorizers of type "request" with `resultTtlInSeconds` set to "0") ([learn more](../deprecations.md#default-identitysource-for-httpauthorizer)).
```
Comment (medikoo) [inline docs/guides/upgrading-v3.md:146]: Is `’` intentional after `AWS`?

**Comment by medikoo** [comment]
> I.e. as a user, reading the upgrade guide, what should I do? How do I know what instruction to follow?

If user decided to install standalone, then should follow its instructions. Upgrade instructions are dedicated for standalone not for any other type of installations

If user is relying on npm, then doesn't need any specific instructions from us (aside of pointing that Framework is available in npm registry, and can be installed via `npm install -g serverless`)

File: docs/guides/upgrading-v3.md (hunk 1 above)
Comment (medikoo) [reply docs/guides/upgrading-v3.md:43]: > I.e. as a user, reading the upgrade guide, what should I do? How do I know what instructions to follow?

If user decided to install a standalone, then should follow its instructions. Upgrade instructions are dedicated for standalone, not for any other type of installations

If user is relying on npm, then doesn't need any specific instructions from us (aside from pointing that Framework is available in npm registry, and can be installed via `npm install -g serverless`)

File: docs/guides/upgrading-v3.md (hunk 8 above)
Comment (mnapoli) [reply docs/guides/upgrading-v3.md]: It seems it depends on whether "AWS" is plural or not 😄 ("services"? or "AWS" as a whole?)  (https://www.thepunctuationguide.com/apostrophe.html)

Anyway if it looks weird that way I might as well go with the "less weird" approach if both could fit.

File: docs/guides/upgrading-v3.md (hunk 1 above)
Comment (mnapoli) [reply docs/guides/upgrading-v3.md:43]: As a user, how can I know how my CLI is installed? (if I installed my CLI 6 months ago, I don't remember)

> If user is relying on npm, then doesn't need any specific instructions from us

They do, some people don't know how to use NPM by heart.

File: docs/guides/upgrading-v3.md (hunk 8 above)
Comment (medikoo) [reply docs/guides/upgrading-v3.md]: I think I did not approach such notation before, but maybe I was simply ignoring it :)

File: docs/guides/upgrading-v3.md (hunk 1 above)
Comment (medikoo) [reply docs/guides/upgrading-v3.md:43]: > They do, some people don't know how to use NPM by heart.

Sure, we may add some more comprehensive section on how to deal with npm, but I would split those two sections, so _npm_ and _standalone_ are not mixed. Those are very different kind of installations. _standalone_ is where we setup _registry_ rules and provide an interface to install and update, so we need to document it.

While in case of _npm_, we simply publish to npm registry, and to consume that users may rely on different PM's. It can be npm, Yarn or PnP etc. Usually projects in their documentation just signal that package is available at npm registry, and having that users rely on instructions of PM of their choice on how to install and update.

Comment (medikoo) [reply docs/guides/upgrading-v3.md:43]: I think if user is not familiar with npm, then should just rely on _standalone_ installation.

IIRC the idea behind current instructions was to favor _standalone_ installation, and on a side, inform that _package_ is available in npm registry for those preferring that way (because it's what most developers are familiar with).

So in light of that, I think there's no real point in providing a comprehensive npm installation doc

Comment (mnapoli) [reply docs/guides/upgrading-v3.md]: As discussed, I've pushed an update where NPM is the default method (I slightly updated the "Getting started" guide too).

I updated the upgrade guide too.

Let me know @medikoo & @pgrzesik

**Review by pgrzesik — APPROVED** [review]
The updated install/upgrade section looks good to me :+1:

**Review by pgrzesik — APPROVED** [review]
:+1:

File: docs/getting-started.md [hunk 9]
```diff
@@ -71,22 +73,22 @@ _Note: Users in China are presented with a setup centered around the chinese [Te
...
+serverless upgrade
 ```
 
-### via npm
+### Windows
 
 ```bash
-npm update -g serverless
+choco upgrade serverless
 ```
```
Comment (medikoo) [inline docs/getting-started.md:92]: It'll be nice to wrap it into dedicated Standalone section as:

```
## Upgrade

### npm installation
...

### Standalone binary

#### macOS/Linux
....

#### Windows
....
```

File: docs/guides/upgrading-v3.md [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+
+```bash
+npm update -g serverless
+```
+
+If you [installed `serverless` as a standalone binary](../getting-started.md#install-as-a-standalone-binary), run the following command instead:
+
+- MacOS/Linux standalone binary: `serverless upgrade --major`
+- Windows: `choco upgrade serverless`
+
```
Comment (medikoo) [inline docs/guides/upgrading-v3.md:51]: Looks great 👍

**Comment by pgrzesik** [comment]
It looks like the validation failed here with too long commit @mnapoli

**Review by medikoo — DISMISSED** [review]
👍

**Comment by mnapoli** [comment]
😢

**Review by medikoo — APPROVED** [review]

**Review by pgrzesik — APPROVED** [review]
:+1:
